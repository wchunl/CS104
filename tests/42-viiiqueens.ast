ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_VARDECL "=" (7.6.18) variable lval
|   |   TOK_ARRAY "array" (7.6.1) array
|   |   |   TOK_TYPE_ID "int" (7.6.7) typeid
|   |   TOK_IDENT "board" (7.6.12) variable
|   |   TOK_NULLPTR "nullptr" (7.6.20) null const
|   TOK_FUNCTION "" (7.8.1) function
|   |   TOK_TYPE_ID "" (7.8.1) typeid
|   |   |   TOK_INT "int" (7.8.1) int
|   |   |   TOK_IDENT "is_safe" (7.8.5) variable
|   |   TOK_PARAM "(" (7.8.13) param
|   |   |   TOK_TYPE_ID "" (7.8.14) typeid
|   |   |   |   TOK_INT "int" (7.8.14) int
|   |   |   |   TOK_IDENT "newcol" (7.8.18) int variable (7.8.14)
|   |   TOK_BLOCK "{" (7.8.26)
|   |   |   TOK_VARDECL "=" (7.9.12) variable lval
|   |   |   |   TOK_INT "int" (7.9.4) int
|   |   |   |   TOK_IDENT "col" (7.9.8) int variable (7.9.12)
|   |   |   |   TOK_INTCON "0" (7.9.14) int const
|   |   |   TOK_VARDECL "=" (7.10.17) variable lval
|   |   |   |   TOK_INT "int" (7.10.4) int
|   |   |   |   TOK_IDENT "diagonal" (7.10.8) int variable (7.10.17)
|   |   |   |   TOK_INTCON "0" (7.10.19) int const
|   |   |   TOK_WHILE "while" (7.11.4)
|   |   |   |   '<' "<" (7.11.15)
|   |   |   |   |   TOK_IDENT "col" (7.11.11) int variable (7.9.12)
|   |   |   |   |   TOK_IDENT "newcol" (7.11.17) int variable (7.8.14)
|   |   |   |   TOK_BLOCK "{" (7.11.25)
|   |   |   |   |   TOK_IF "if" (7.12.7)
|   |   |   |   |   |   TOK_EQ "==" (7.12.22) vreg
|   |   |   |   |   |   |   TOK_INDEX "[" (7.12.16) lval vaddr
|   |   |   |   |   |   |   |   TOK_IDENT "board" (7.12.11) variable
|   |   |   |   |   |   |   |   TOK_IDENT "col" (7.12.17) int variable (7.9.12)
|   |   |   |   |   |   |   TOK_INDEX "[" (7.12.30) lval vaddr
|   |   |   |   |   |   |   |   TOK_IDENT "board" (7.12.25) variable
|   |   |   |   |   |   |   |   TOK_IDENT "newcol" (7.12.31) int variable (7.8.14)
|   |   |   |   |   |   TOK_RETURN "return" (7.12.40)
|   |   |   |   |   |   |   TOK_INTCON "0" (7.12.47) int const
|   |   |   |   |   '=' "=" (7.13.16) int vreg
|   |   |   |   |   |   TOK_IDENT "diagonal" (7.13.7) int variable (7.10.17)
|   |   |   |   |   |   '-' "-" (7.13.29) int vreg
|   |   |   |   |   |   |   TOK_INDEX "[" (7.13.23) lval vaddr
|   |   |   |   |   |   |   |   TOK_IDENT "board" (7.13.18) variable
|   |   |   |   |   |   |   |   TOK_IDENT "col" (7.13.24) int variable (7.9.12)
|   |   |   |   |   |   |   TOK_INDEX "[" (7.13.36) lval vaddr
|   |   |   |   |   |   |   |   TOK_IDENT "board" (7.13.31) variable
|   |   |   |   |   |   |   |   TOK_IDENT "newcol" (7.13.37) int variable (7.8.14)
|   |   |   |   |   TOK_IF "if" (7.14.7)
|   |   |   |   |   |   TOK_EQ "==" (7.14.20) vreg
|   |   |   |   |   |   |   TOK_IDENT "diagonal" (7.14.11) int variable (7.10.17)
|   |   |   |   |   |   |   '-' "-" (7.14.27) int vreg
|   |   |   |   |   |   |   |   TOK_IDENT "col" (7.14.23) int variable (7.9.12)
|   |   |   |   |   |   |   |   TOK_IDENT "newcol" (7.14.29) int variable (7.8.14)
|   |   |   |   |   |   TOK_RETURN "return" (7.14.37)
|   |   |   |   |   |   |   TOK_INTCON "0" (7.14.44) int const
|   |   |   |   |   TOK_IF "if" (7.15.7)
|   |   |   |   |   |   TOK_EQ "==" (7.15.20) vreg
|   |   |   |   |   |   |   TOK_IDENT "diagonal" (7.15.11) int variable (7.10.17)
|   |   |   |   |   |   |   '-' "-" (7.15.30) int vreg
|   |   |   |   |   |   |   |   TOK_IDENT "newcol" (7.15.23) int variable (7.8.14)
|   |   |   |   |   |   |   |   TOK_IDENT "col" (7.15.32) int variable (7.9.12)
|   |   |   |   |   |   TOK_RETURN "return" (7.15.37)
|   |   |   |   |   |   |   TOK_INTCON "0" (7.15.44) int const
|   |   |   |   |   '=' "=" (7.16.11) int vreg
|   |   |   |   |   |   TOK_IDENT "col" (7.16.7) int variable (7.9.12)
|   |   |   |   |   |   '+' "+" (7.16.17) int vreg
|   |   |   |   |   |   |   TOK_IDENT "col" (7.16.13) int variable (7.9.12)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.16.19) int const
|   |   |   TOK_RETURN "return" (7.18.4)
|   |   |   |   TOK_INTCON "1" (7.18.11) int const
|   TOK_FUNCTION "" (7.21.1) function
|   |   TOK_TYPE_ID "" (7.21.1) typeid
|   |   |   TOK_VOID "void" (7.21.1) void
|   |   |   TOK_IDENT "printqueens" (7.21.6) variable
|   |   TOK_PARAM "(" (7.21.17) param
|   |   TOK_BLOCK "{" (7.21.20)
|   |   |   TOK_VARDECL "=" (7.22.12) variable lval
|   |   |   |   TOK_INT "int" (7.22.4) int
|   |   |   |   TOK_IDENT "col" (7.22.8) int variable (7.9.12)
|   |   |   |   TOK_INTCON "0" (7.22.14) int const
|   |   |   TOK_WHILE "while" (7.23.4)
|   |   |   |   '<' "<" (7.23.15)
|   |   |   |   |   TOK_IDENT "col" (7.23.11) int variable (7.9.12)
|   |   |   |   |   TOK_INTCON "8" (7.23.17) int const
|   |   |   |   TOK_BLOCK "{" (7.23.20)
|   |   |   |   |   TOK_CALL "(" (7.24.14)
|   |   |   |   |   |   TOK_IDENT "putchr" (7.24.7) variable
|   |   |   |   |   |   '+' "+" (7.24.26) int vreg
|   |   |   |   |   |   |   TOK_INDEX "[" (7.24.20) lval vaddr
|   |   |   |   |   |   |   |   TOK_IDENT "board" (7.24.15) variable
|   |   |   |   |   |   |   |   TOK_IDENT "col" (7.24.21) int variable (7.9.12)
|   |   |   |   |   |   |   TOK_CHARCON "'1'" (7.24.28) int const
|   |   |   |   |   '=' "=" (7.25.11) int vreg
|   |   |   |   |   |   TOK_IDENT "col" (7.25.7) int variable (7.9.12)
|   |   |   |   |   |   '+' "+" (7.25.17) int vreg
|   |   |   |   |   |   |   TOK_IDENT "col" (7.25.13) int variable (7.9.12)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.25.19) int const
|   |   |   TOK_CALL "(" (7.27.11)
|   |   |   |   TOK_IDENT "putchr" (7.27.4) variable
|   |   |   |   TOK_CHARCON "'\n'" (7.27.12) int const
|   TOK_FUNCTION "" (7.30.1) function
|   |   TOK_TYPE_ID "" (7.30.1) typeid
|   |   |   TOK_VOID "void" (7.30.1) void
|   |   |   TOK_IDENT "queens" (7.30.6) variable
|   |   TOK_PARAM "(" (7.30.13) param
|   |   |   TOK_TYPE_ID "" (7.30.14) typeid
|   |   |   |   TOK_INT "int" (7.30.14) int
|   |   |   |   TOK_IDENT "newcol" (7.30.18) int variable (7.8.14)
|   |   TOK_BLOCK "{" (7.30.26)
|   |   |   TOK_VARDECL "=" (7.31.12) variable lval
|   |   |   |   TOK_INT "int" (7.31.4) int
|   |   |   |   TOK_IDENT "row" (7.31.8) int variable (7.31.12)
|   |   |   |   TOK_INTCON "0" (7.31.14) int const
|   |   |   TOK_IF "if" (7.32.4)
|   |   |   |   TOK_EQ "==" (7.32.15) int vreg
|   |   |   |   |   TOK_IDENT "newcol" (7.32.8) int variable (7.8.14)
|   |   |   |   |   TOK_INTCON "8" (7.32.18) int const
|   |   |   |   TOK_CALL "(" (7.32.32)
|   |   |   |   |   TOK_IDENT "printqueens" (7.32.21) variable
|   |   |   |   TOK_IFELSE "{" (7.33.9)
|   |   |   |   |   TOK_WHILE "while" (7.34.7)
|   |   |   |   |   |   '<' "<" (7.34.18)
|   |   |   |   |   |   |   TOK_IDENT "row" (7.34.14) int variable (7.31.12)
|   |   |   |   |   |   |   TOK_INTCON "8" (7.34.20) int const
|   |   |   |   |   |   TOK_BLOCK "{" (7.34.23)
|   |   |   |   |   |   |   '=' "=" (7.35.24) int vreg
|   |   |   |   |   |   |   |   TOK_INDEX "[" (7.35.15) lval vaddr
|   |   |   |   |   |   |   |   |   TOK_IDENT "board" (7.35.10) variable
|   |   |   |   |   |   |   |   |   TOK_IDENT "newcol" (7.35.16) int variable (7.8.14)
|   |   |   |   |   |   |   |   TOK_IDENT "row" (7.35.26) int variable (7.31.12)
|   |   |   |   |   |   |   TOK_IF "if" (7.36.10)
|   |   |   |   |   |   |   |   TOK_CALL "(" (7.36.22)
|   |   |   |   |   |   |   |   |   TOK_IDENT "is_safe" (7.36.14) variable
|   |   |   |   |   |   |   |   |   TOK_IDENT "newcol" (7.36.23) int variable (7.8.14)
|   |   |   |   |   |   |   |   TOK_CALL "(" (7.36.39)
|   |   |   |   |   |   |   |   |   TOK_IDENT "queens" (7.36.32) variable
|   |   |   |   |   |   |   |   |   '+' "+" (7.36.47) int vreg
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "newcol" (7.36.40) int variable (7.8.14)
|   |   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.36.49) int const
|   |   |   |   |   |   |   '=' "=" (7.37.14) int vreg
|   |   |   |   |   |   |   |   TOK_IDENT "row" (7.37.10) int variable (7.31.12)
|   |   |   |   |   |   |   |   '+' "+" (7.37.20) int vreg
|   |   |   |   |   |   |   |   |   TOK_IDENT "row" (7.37.16) int variable (7.31.12)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.37.22) int const
|   TOK_FUNCTION "" (7.42.1) function
|   |   TOK_TYPE_ID "" (7.42.1) typeid
|   |   |   TOK_INT "int" (7.42.1) int
|   |   |   TOK_IDENT "main" (7.42.5) variable
|   |   TOK_PARAM "(" (7.42.9) param
|   |   TOK_BLOCK "{" (7.42.12)
|   |   |   '=' "=" (7.43.10) int vreg
|   |   |   |   TOK_IDENT "board" (7.43.4) variable
|   |   |   |   TOK_ALLOC "alloc" (7.43.12)
|   |   |   |   |   TOK_ARRAY "array" (7.43.18) array
|   |   |   |   |   TOK_TYPE_ID "int" (7.43.24) typeid
|   |   |   |   |   TOK_INTCON "8" (7.43.31) int const
|   |   |   TOK_CALL "(" (7.44.11)
|   |   |   |   TOK_IDENT "queens" (7.44.4) variable
|   |   |   |   TOK_INTCON "0" (7.44.12) int const
|   |   |   TOK_RETURN "return" (7.45.4)
|   |   |   |   TOK_INTCON "0" (7.45.11) int const
