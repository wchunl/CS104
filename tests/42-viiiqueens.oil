fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
          board:.global ptr nullptr
is_safe:  .function int
          .param int newcol
          .local int col
          .local int diagonal
          col = 0
          diagonal = 0
.wh0:     $t0:i = col < newcol
          goto .od0 if not $t0:i
.do0:     
.if0:     $t1:p = board[newcol * :p]
          $t2:i = $t1:i == [
          goto .el0 if not $t2:i
.th0:     
.fi0:     $t3:p = board[newcol * :p]
          $t4:i = $t3:i - [
          diagonal = $t4:i
.if1:     $t5:i = col - newcol
          $t6:i = $t5:i == -
          goto .el1 if not $t6:i
.th1:     
.fi1:     
.if2:     $t7:i = newcol - col
          $t8:i = $t7:i == -
          goto .el2 if not $t8:i
.th2:     
.fi2:     $t9:i = col + 1
          col = $t9:i
.od0:     return 1
          return 
          .end
printqueens:.function
          .local int col
          col = 0
.wh1:     $t10:i = col < 8
          goto .od1 if not $t10:i
.do1:     $t11:i = board[col * :p]
          $t12:i = $t11:i + '1'
          putchr($t12:i)
          $t13:i = col + 1
          col = $t13:i
.od1:     putchr('\n')
          .end
queens:   .function
          .param int newcol
          .local int row
          row = 0
.if3:     $t14:i = newcol == 8
          goto .el3 if not $t14:i
.th3:     printqueens()
.el3:     
.wh2:     $t15:i = row < 8
          goto .od2 if not $t15:i
.do2:     board[newcol * :p] = row
.if4:     is_safe(newcol)
          goto .el4 if not $t16:i
.th4:     $t17:i = newcol + 1
          queens($t17:i)
.fi4:     $t18:i = row + 1
          row = $t18:i
.od2:     
.fi3:     .end
main:     .function int
          malloc(4 * sizeof int)
          board = $t19:i
          queens(0)
          return 0
          return 
          .end
