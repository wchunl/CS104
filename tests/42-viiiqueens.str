string_set[   2]:   12597418120160561823 0x1650d18->"int"
string_set[   3]:   15403598725584096552 0x1653328->"diagonal"
                   18004526203352653430 0x1652cc8->"is_safe"
string_set[   5]:    1370730858159036685 0x1654a78->"+"
string_set[   8]:   13497578857207247412 0x1650bb8->"string"
string_set[   9]:   10319632053171904911 0x1653348->"while"
string_set[  10]:    9201800968376702500 0x1651fa8->"getchr"
string_set[  14]:   15816506932648033945 0x164c328->"("
                    1396094175042499165 0x16528a8->"<"
string_set[  15]:   12868445110721718657 0x1650f18->")"
string_set[  17]:    6142509188972423790 0x164c6d8->""
                    4132753638312158367 0x1652ca8->"newcol"
string_set[  20]:    1583576014053361350 0x1651b78->"str"
string_set[  21]:    9317282556992397945 0x1653548->"if"
string_set[  22]:   10159970873491820195 0x1654ba8->"1"
string_set[  26]:   12291392672316406109 0x16510d8->"line"
                    9611428142580219387 0x164c6f8->"fail"
                   11597697714117577063 0x1652b48->"="
string_set[  28]:   18413295556186398326 0x16512e8->"putint"
string_set[  29]:   11911796782241914524 0x16507d8->"void"
string_set[  31]:    8475192774821045282 0x16554c8->"'1'"
string_set[  33]:    6814028903552918164 0x1652718->"array"
string_set[  35]:   11868120863286193964 0x1657438->"main"
string_set[  41]:   12051031859636400188 0x1653b08->"]"
string_set[  42]:   13686721134187162920 0x1654c58->"}"
string_set[  49]:    7574291458008592867 0x16529c8->"queens"
                   12679825168387489523 0x16519c8->"num"
string_set[  50]:      21606623943142546 0x1653058->"{"
string_set[  52]:    5344662657899890615 0x16510f8->";"
string_set[  53]:   10836626480094982265 0x1653168->"col"
string_set[  54]:    7296690267450279874 0x16515a8->"chr"
string_set[  55]:    2297668033614959926 0x1653278->"0"
string_set[  59]:    6192799629375248351 0x1655268->"'\n'"
string_set[  60]:    1873634190020599143 0x1656768->"else"
string_set[  63]:   14526079891305879776 0x1651288->">"
string_set[  64]:   16579916199511856337 0x16524a8->"getln"
string_set[  65]:    7110316202799739443 0x1652af8->"board"
string_set[  66]:   17899137375351314596 0x1654d08->"8"
string_set[  67]:    4706206694200543916 0x1650de8->","
                   12808232827189627082 0x1652bf8->"nullptr"
string_set[  73]:    9386170080689500517 0x1650c98->"expr"
string_set[  76]:   11166701219533395393 0x1652218->"getstr"
string_set[  77]:   11526621952165666095 0x1654d98->"printqueens"
string_set[  80]:    8290819294854170237 0x16563f8->"alloc"
                    7211130878751977441 0x1651438->"putchr"
string_set[  83]:   13581035385482130682 0x16539f8->"["
string_set[  85]:    4017852018407011111 0x1650ef8->"file"
                    7372855250431704371 0x1654078->"-"
string_set[  87]:   14635384904416252381 0x1653d48->"return"
string_set[  89]:    8374522807849457143 0x1656268->"row"
                   13348923631108115385 0x1651758->"putstr"
string_set[  95]:   14736352883882276958 0x1653b28->"=="
load_factor = 0.536
bucket_count = 97
max_bucket_size = 3
