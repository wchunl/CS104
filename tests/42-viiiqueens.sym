fail (6.22.1) {0} void function
   expr (6.22.12) {1} string variable param lval 0
   file (6.22.25) {1} string variable param lval 1
   line (6.22.38) {1} int variable param lval 2

putchr (6.24.1) {1} void function
   chr (6.24.14) {1} int variable param lval 0

putint (6.25.1) {2} void function
   num (6.25.14) {1} int variable param lval 0

putstr (6.26.1) {3} void function
   str (6.26.14) {1} string variable param lval 0

getchr (6.28.1) {4} int function

getstr (6.29.1) {5} string function

getln (6.30.1) {6} string function

board (7.6.1) array <int> variable lval 0

is_safe (7.8.1) {7} int function
   newcol (7.8.14) {1} int variable param lval 0
   col (7.9.12) {1} int variable local lval 0
   diagonal (7.10.17) {1} int variable local lval 1

printqueens (7.21.1) {8} void function
   col (7.22.12) {1} int variable local lval 0

queens (7.30.1) {9} void function
   newcol (7.30.14) {1} int variable param lval 0
   row (7.31.12) {1} int variable local lval 0

main (7.42.1) {10} int function
