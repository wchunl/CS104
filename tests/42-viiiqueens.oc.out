; 0.  1: # 1 "42-viiiqueens.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "42-viiiqueens.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "42-viiiqueens.oc" 2
; 7.  4: 
; 7.  5: 
; 7.  6: array<int> board = nullptr;
; 7.  7: 
; 7.  8: int is_safe (int newcol) {
; 7.  9:    int col = 0;
; 7. 10:    int diagonal = 0;
; 7. 11:    while (col < newcol) {
; 7. 12:       if (board[col] == board[newcol]) return 0;
; 7. 13:       diagonal = board[col] - board[newcol];
; 7. 14:       if (diagonal == col - newcol) return 0;
; 7. 15:       if (diagonal == newcol - col) return 0;
; 7. 16:       col = col + 1;
; 7. 17:    }
; 7. 18:    return 1;
; 7. 19: }
; 7. 20: 
; 7. 21: void printqueens() {
; 7. 22:    int col = 0;
; 7. 23:    while (col < 8) {
; 7. 24:       putchr (board[col] + '1');
; 7. 25:       col = col + 1;
; 7. 26:    }
; 7. 27:    putchr ('\n');
; 7. 28: }
; 7. 29: 
; 7. 30: void queens (int newcol) {
; 7. 31:    int row = 0;
; 7. 32:    if (newcol == 8) printqueens();
; 7. 33:    else {
; 7. 34:       while (row < 8) {
; 7. 35:          board[newcol] = row;
; 7. 36:          if (is_safe (newcol)) queens (newcol + 1);
; 7. 37:          row = row + 1;
; 7. 38:       }
; 7. 39:    }
; 7. 40: }
; 7. 41: 
; 7. 42: int main() {
; 7. 43:    board = alloc<array<int>> (8);
; 7. 44:    queens (0);
; 7. 45:    return 0;
; 7. 46: }
